File: leetcode/word-search/word-search.cpp
/**************************************
* @author dujiong
* @date 2016.12.22
* @version V0.1
**************************************/

class Solution {
public:
    bool exist(vector<vector<char> > &board, string word) {
        if(board.size()==0) return false;
        if(word.length()==0) return false;
        int i,j; 
        N=board.size(); 
        M=board[0].size();
        vector<vector<int> > book(N,vector<int>(M,0));
        for(i=0;i<N;i++)
            for(j=0;j<M;j++)
                if(dfs(board,i,j,word,0,book)==true) return true;
        return false;
    }
private:
    int N,M;
    bool dfs(vector<vector<char> > &board,int x,int y,string word,int p,vector<vector<int> > &book){
        if(p==word.length()) return true;
        if(x<0||x>=N||y<0||y>=M) return false;
        if(book[x][y]==1) return false;
        if(board[x][y]!=word[p]) return false;
        book[x][y]=1;
        bool k=dfs(board,x+1,y,word,p+1,book)||dfs(board,x-1,y,word,p+1,book)||dfs(board,x,y+1,word,p+1,book)||dfs(board,x,y-1,word,p+1,book);
        book[x][y]=0;
        return k;
    }
};

File: leetcode-golang/Word Search/Word Search.go
func exist(board [][]byte, word string) bool {
	if len(board) == 0 {
		return true
	}
	visited := make([][]bool, len(board))
	for i := 0; i < len(board); i++ {
		visited[i] = make([]bool, len(board[0]))
	}
	for i := 0; i < len(board); i++ {
		for j := 0; j < len(board[0]); j++ {
			if dfs(board, i, j, 0, visited, word) { //这里不能直接return dfs(...)，因为对于i,j很多种情况，这样的话只会是i=0,j=0的情况
				return true
			}
		}
	}
	return false
}

func dfs(board [][]byte, i, j int, index int, visited [][]bool, word string) bool {
	if index == len(word) {
		return true
	}
	if i < 0 || j < 0 || i >= len(board) || j >= len(board[0]) || visited[i][j] || board[i][j] != word[index] {
		return false
	}
	visited[i][j] = true
	res := dfs(board, i+1, j, index+1, visited, word) ||
		dfs(board, i-1, j, index+1, visited, word) ||
		dfs(board, i, j+1, index+1, visited, word) ||
		dfs(board, i, j-1, index+1, visited, word)
	visited[i][j] = false
	return res
}
